### PR TITLE
fix(hex): handle empty helper output and fix public key response tuple order

### PR DESCRIPTION
**Pull request by thavaahariharangit**
### What are you trying to accomplish?

Fixing https://github.com/dependabot/dependabot-core/issues/14168 - Hex builds fail when fetching the public key from a private registry.

**Root cause:** Two bugs in the Elixir helper script (run.exs):
- `fetch_public_key` destructured the `Hex.Repo.get_public_key` response tuple as `{status, body, headers}` when the actual shape is `{status, headers, body}`, so the public key was never extracted correctly.
- When the helper subprocess returned empty output (e.g., a registry returning no data), `binary_to_term("")` crashed instead of producing a clear error.

Changes:
- `run.exs`: Fix `fetch_public_key` tuple pattern from `{200, key, _}` to `{200, _headers, key}`. Add an empty-output guard that returns a descriptive error instead of crashing on `binary_to_term`.
- `lockfile_updater_spec.rb`: Add test coverage for public key fingerprint mismatch, public key download failure, and empty helper output error paths.
- `update_checker_spec.rb`: Add test coverage for registry verification crash (empty output) and public key download HTTP error scenarios.

### How will you know you've accomplished your goal?

1. **All existing Hex tests pass** — no regressions introduced.
2. **New tests pass** for the added error paths: public key fingerprint mismatch, public key download failure, and empty helper output — in both `lockfile_updater_spec.rb` and `update_checker_spec.rb`.
3. **Manual verification** against a JFrog private Hex registry: when the registry returns a non-200 status for the public key endpoint, the updater now raises a `PrivateSourceAuthenticationFailure` with the repo name instead of crashing with an opaque `binary_to_term` error.

```
updater | 2026/02/18 09:52:56 INFO Started process PID: 4831 with command: {"MIX_EXS" => "/opt/hex/mix.exs", "MIX_QUIET" => "1"} mix run /opt/hex/lib/run.exs
  proxy | 2026/02/18 09:52:58 [014] GET https://jfrogghdemo.jfrog.io:443/artifactory/api/hex/dependabot_hex/public_key
  proxy | 2026/02/18 09:52:59 [014] 501 https://jfrogghdemo.jfrog.io:443/artifactory/api/hex/dependabot_hex/public_key
```

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: hex/spec/dependabot/hex/update_checker_spec.rb [hunk 1]
```diff
@@ -707,6 +707,68 @@
...
+              )
+            )
+        end
+
+        it "raises a helpful error" do
+          expect { latest_resolvable_version }
+            .to raise_error(Dependabot::DependencyFileNotResolvable) do |error|
+              expect(error.message).to include("No output returned from helper script")
+            end
+        end
```
Comment (robaiken) [inline hex/spec/dependabot/hex/update_checker_spec.rb]: I'm struggling to see the value these tests actually provide. The way they're set up, you could pass any message to `HelperSubprocessFailed` and just assert that same message comes back out. For example you could set the message to `Hello, world!` and then check that `error.message` includes `Hello, world!` and the test would pass. There are too many things being mocked here so we end up only testing the mock itself rather than any real behaviour

File: hex/helpers/lib/run.exs [hunk 2]
```diff
@@ -121,7 +124,7 @@ defmodule DependencyHelper do
 
   defp fetch_public_key(repo, repo_url, auth_key, fingerprint) do
     case Hex.Repo.get_public_key(%{trusted: true, url: repo_url, auth_key: auth_key}) do
-      {:ok, {200, key, _}} ->
+      {:ok, {200, _headers, key}} ->
```
Comment (robaiken) [inline hex/helpers/lib/run.exs:127]: I agree than this change needs to be made based on the [hex docs](https://hexdocs.pm/hex_core/hex_repo.html#get_public_key/1). But I believe that this will only fix this affect self hosted runners and its possible that you will have to update [Dependabot proxy](https://github.com/dependabot/proxy) as well

File: hex/spec/dependabot/hex/update_checker_spec.rb (hunk 1 above)
Comment (v-thavaahariharangit) [reply hex/spec/dependabot/hex/update_checker_spec.rb]: Thanks, you're right that these tests were only verifying the mock itself rather than any real behavior. I've removed both test contexts ("when registry verification crashes with no output" and "when public key download returns HTTP error code") from `update_checker_spec.rb` in the latest push. The error-handling regex logic is already properly covered by the equivalent tests in `lockfile_updater_spec.rb`.

File: hex/helpers/lib/run.exs (hunk 2 above)
Comment (v-thavaahariharangit) [reply hex/helpers/lib/run.exs:127]: Thanks for confirming the tuple order fix is correct per the hex docs. Good call on the Dependabot proxy — I'll take a look at the [proxy repo](https://github.com/dependabot/proxy) to check whether it also needs a corresponding update for this to work end-to-end on hosted runners.

**Comment by kybishop** [comment]
Is there anything holding this PR up? Dependabot has been broken for a staggering number of elixir projects for over a month now 🫠

**Comment by v-thavaahariharangit** [comment]
covered here: https://github.com/dependabot/dependabot-core/pull/14380
